src/System.Text.Json/src/System/Text/Json/Reader/JsonReaderHelper.Unescaping.cs

```diff
@@ -0,0 +1,210 @@
+        }
+
+        /// <summary>
+        /// Copies the UTF-8 code unit representation of this scalar to an output buffer.
+        /// The buffer must be large enough to hold the required number of <see cref="byte"/>s.
+        /// </summary>
+        private static void EncodeToUtf8Bytes(uint scalar, Span<byte> utf8Destination, out int bytesWritten)
+        {
+            Debug.Assert(IsValidUnicodeScalar(scalar));
+            Debug.Assert(utf8Destination.Length >= 4);
+
+            if (scalar < 0x80U)
+            {
+                // Single UTF-8 code unit
+                utf8Destination[0] = (byte)scalar;
+                bytesWritten = 1;
+            }
+            else if (scalar < 0x800U)
+            {
+                // Two UTF-8 code units
+                utf8Destination[0] = (byte)(0xC0U | (scalar >> 6));
+                utf8Destination[1] = (byte)(0x80U | (scalar & 0x3FU));
+                bytesWritten = 2;
+            }
+            else if (scalar < 0x10000U)
+            {
+                // Three UTF-8 code units
+                utf8Destination[0] = (byte)(0xE0U | (scalar >> 12));
+                utf8Destination[1] = (byte)(0x80U | ((scalar >> 6) & 0x3FU));
+                utf8Destination[2] = (byte)(0x80U | (scalar & 0x3FU));
+                bytesWritten = 3;
+            }
+            else
+            {
+                // Four UTF-8 code units
+                utf8Destination[0] = (byte)(0xF0U | (scalar >> 18));
+                utf8Destination[1] = (byte)(0x80U | ((scalar >> 12) & 0x3FU));
+                utf8Destination[2] = (byte)(0x80U | ((scalar >> 6) & 0x3FU));
+                utf8Destination[3] = (byte)(0x80U | (scalar & 0x3FU));
+                bytesWritten = 4;
+            }
+        }
+
+        /// <summary>
+        /// Returns <see langword="true"/> iff <paramref name="value"/> is a valid Unicode scalar
+        /// value, i.e., is in [ U+0000..U+D7FF ], inclusive; or [ U+E000..U+10FFFF ], inclusive.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsValidUnicodeScalar(uint value)
+        {
+            // By XORing the incoming value with 0xD800, surrogate code points
+            // are moved to the range [ U+0000..U+07FF ], and all valid scalar
+            // values are clustered into the single range [ U+0800..U+10FFFF ],
+            // which allows performing a single fast range check.
+
+            return IsInRangeInclusive(value ^ 0xD800U, 0x800U, 0x10FFFFU);
+        }
+
+        /// <summary>
+        /// Returns <see langword="true"/> iff <paramref name="value"/> is between
+        /// <paramref name="lowerBound"/> and <paramref name="upperBound"/>, inclusive.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsInRangeInclusive(uint value, uint lowerBound, uint upperBound)
+            => (value - lowerBound) <= (upperBound - lowerBound);
+    }
+}
```

Comment:
We're writing all this code out manually because we're not able to use Rune for the source build?  If that's the reason, I'd rather not have all of this code duplicated for the binary build; can it be put into a partial file with Rune-based implementations on .NET Core and manually written out for the source build?

Reply:
These APIs are needed by the writer escaping logic already, so the only one that I can move is `EncodeToUtf8Bytes`